beacon_node/lighthouse_network/src/rpc/active_requests_limiter.rs

```diff
@@ -0,0 +1,178 @@
+use crate::rpc::{Protocol, SubstreamId};
+use libp2p::swarm::ConnectionId;
+use libp2p::PeerId;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::time::Duration;
+use tokio::time::Instant;
+
+/// Restricts more than two requests from running simultaneously on the same protocol per peer.
+pub(super) struct ActiveRequestsLimiter {
+    resp_timeout: Duration,
+    requests: HashMap<PeerId, Vec<(Protocol, ConnectionId, SubstreamId, Instant)>>,
```

Comment:
I'm not convinced why we need `ConnectionId` and `SubstreamId` here. 
They would be relevant in the case where a peer has multiple connections but even in that case, I don't think we want a peer to split multiple requests over multiple connection ids.

Even for `Instant`, I'm not fully convinced. We wouldn't respond within 15 secs only when we are heavily overloaded so we are probably not capable of serving a new request and in this case. Are there any other cases where replacing the request would be relevant?

Adding `Instant` makes sense but doesn't seem to useful to me in a realistic scenario.

Reply:
> Therefore, the behaviour can't then remove the Request from the ActiveRequestsLimiter
as we currently don't have Protocol on the RPCCodedResponse, 

Yes, that's why I think we need the combination of `ConnectionId` and `SubstreamId`.

> but imho it's cleaner to just have HashSet<(PeerId, Protocol)> and align on a single Protocol enum in the repo (for now we can have From<&RPCCodedResponse> for Protocol)

That would be cleaner, but I'm not sure if we can do that, because `RPCCodedResponse::Error` doesn’t contain a protocol. 🤔

Reply:
> Even for Instant, I'm not fully convinced. We wouldn't respond within 15 secs only when we are heavily overloaded so we are probably not capable of serving a new request and in this case. Are there any other cases where replacing the request would be relevant?

The intention of adding `Instant` was to ensure this limiter allows new requests even if there are some bugs that cause requests to remain in this limiter despite the response having been sent.

I thought again about this, and yeah, it might not be useful. I’m happy to remove it. 🙂

Reply:
Removed `Instant` in  https://github.com/sigp/lighthouse/pull/5923/commits/a325438d8ca141b71501af44ac25e54f83f8f182